folly/detail/TypeList.h

```diff
@@ -168,7 +168,7 @@ using If = MetaApply<impl::If_<If_>, Then, Else>;
  */
 template <template <class...> class C, class... Ts>
 class MetaDefer {
-  template <template <class...> class D = C, class = D<Ts...>>
+  template <class = C<Ts...>>
```

Comment:
Why was this necessary?

Reply:
I was developing with VS2019 Preview 16.2.0 Preview 3.0 and as originally written that line gives the error `error C3202: 'C': invalid default argument, expected a class template`

However I just checked and the original is fine with VS2019 Release 16.1.5 so I will report a bug to the VS team immediately.

I think it creates the same substitution failure written either way though?

Reply:
Wait no sorry it's broken on VS2019 16.1.5 as well. Will report a bug either way as I think it should be valid as written.

Reply:
Seeing this under recent Clang:

```
folly/detail/TypeList.h:171:21: error: too many template arguments for alias template 'AddPointer'
  template <class = C<Ts...>>
                    ^ ~~~~~~
folly/detail/TypeList.h:90:1: note: in instantiation of template class 'folly::detail::MetaDefer<AddPointer, int, short>' requested here
using MetaApply = typename Fn::template apply<Ts...>;
^
folly/test/TypeListTest.cpp:27:49: note: in instantiation of template type alias 'MetaApply' requested here
struct IsApplicable_<T, TypeList<Ts...>, void_t<MetaApply<T, Ts...>>>
                                                ^
folly/test/TypeListTest.cpp:45:18: note: during template argument deduction for class template partial specialization 'IsApplicable_<T, TypeList<Ts...>, void_t<MetaApply<T, Ts...> > >' [with T = folly::detail::MetaDefer<AddPointer, int, short>, Ts = <>]
  static_assert(!IsApplicable<MetaDefer<AddPointer, int, short>>::value, "");
                 ^
folly/test/TypeListTest.cpp:45:18: note: in instantiation of template class '(anonymous namespace)::IsApplicable_<folly::detail::MetaDefer<AddPointer, int, short>, folly::detail::TypeList<>, void>' requested here
folly/test/TypeListTest.cpp:40:1: note: template is declared here
using AddPointer = T*;
^
folly/test/TypeListTest.cpp:40:1: error: 'AddPointer' declared as a pointer to a reference of type 'int &'
using AddPointer = T*;
^
folly/detail/TypeList.h:171:21: note: in instantiation of template type alias 'AddPointer' requested here
  template <class = C<Ts...>>
                    ^
folly/detail/TypeList.h:90:1: note: in instantiation of template class 'folly::detail::MetaDefer<AddPointer, int &>' requested here
using MetaApply = typename Fn::template apply<Ts...>;
^
folly/test/TypeListTest.cpp:27:49: note: in instantiation of template type alias 'MetaApply' requested here
struct IsApplicable_<T, TypeList<Ts...>, void_t<MetaApply<T, Ts...>>>
                                                ^
folly/test/TypeListTest.cpp:46:18: note: during template argument deduction for class template partial specialization 'IsApplicable_<T, TypeList<Ts...>, void_t<MetaApply<T, Ts...> > >' [with T = folly::detail::MetaDefer<AddPointer, int &>, Ts = <>]
  static_assert(!IsApplicable<MetaDefer<AddPointer, int&>>::value, "");
                 ^
folly/test/TypeListTest.cpp:46:18: note: in instantiation of template class '(anonymous namespace)::IsApplicable_<folly::detail::MetaDefer<AddPointer, int &>, folly::detail::TypeList<>, void>' requested here
```

Reply:
Ack good catch. Was not aware of `TypeListTest.cpp` which I see is not added in the root `CMakeLists.txt`. I've added it in my local copy of the repo and tried a new fix. Seems to work locally with MSVC, and my clang-8 and gcc 8.1 installs as well.